src/compiler/factory.ts

```diff
@@ -2888,7 +2895,9 @@ namespace ts {
             templateObjectHelper,
             generatorHelper,
             importStarHelper,
-            importDefaultHelper
+            importDefaultHelper,
+            classPrivateFieldGetHelper,
+            classPrivateFieldSetHelper

```

Comment:
@ExE-Boss cc @Jamesernator @joeywatts @tim-mc

re https://github.com/microsoft/TypeScript/pull/30829/#issuecomment-540070660

> Babel uses property descriptors as the values in the WeakMap, which I feel is more future‑proof.

@ExE-Boss thanks for implementing this in tslib!

How is using property descriptors more future-proof? Are you thinking of a time when decorators will be able to modify private-identified properties?

Before making the change, I'd be curious about the performance impact and readability of the output.

Reply:
also, re @Jamesernator suggesting the change for target >= ES6: we do not downlevel private identifiers for < ES6/ES2015 because we don't know a feasible way of doing it without leaking memory.

Reply:
Isn’t “works but leaks memory” better than “doesn’t work”?

Reply:
@ljharb re
> Isn’t “works but leaks memory” better than “doesn’t work”?

It's an interesting question. If TS team were interested in downleveling to a leaky fake WeakMap semi-polyfill for ES5 and lower, I think it would make sense to do it in a follow-on PR.

I personally think such a downlevel would be risky and would not add value for users, but to each their own.

Reply:
I assume it depends how many objects an app attaches private data to that then otherwise become collectible; if that number is small, the leak would be insignificant.

(I also made https://npmjs.com/internal-slot which falls back cleanly from WeakMap to Map to an array, which is as non-leaky and performant as i could manage without mutating object keys)

Reply:
> Isn’t “works but leaks memory” better than “doesn’t work”?

> (I also made https://npmjs.com/internal-slot which falls back cleanly from WeakMap to Map to an array, which is as non-leaky and performant as i could manage without mutating object keys)

There's two choices for down-leveling, either leak memory or add an additional property onto the object that can be collected with the object however this makes it public.

Both versions of down-leveling have their use case, some people may just wish to use private fields so they don't need to worry about name collisions like with `private fieldName` in subclasses. Others may need strict privacy in which case the leaky variant is more valuable.

I feel the best option would be to add a compiler option `"downlevelPrivateFieldsToES5"` that can be set to either `"useMaps"` or `"usePublicProperty"` depending on what their preference is.

> How is using property descriptors more future-proof? Are you thinking of a time when decorators will be able to modify private-identified properties?

This is probably because there's a proposal for private methods/getters/setters: https://github.com/tc39/proposal-private-methods

Reply:
Can I mark this thread as resolved?

Fwiw, this PR does not permanently settle the story for downleveling to sub-ES2015 targets, so that discussion could be moved to a dedicated issue.

Reply:
Thanks Jordan.

If anyone makes an issue, could you post the link here? I'm curious to follow the discussion and contribute two cents where I can.